File: src/csharp/Intel/Iced/Intel/NasmFormatterInternal/InstrInfos.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if NASM
using System;
using Iced.Intel.FormatterInternal;
using Iced.Intel.Internal;

namespace Iced.Intel.NasmFormatterInternal {
	static partial class InstrInfos {
		public static readonly InstrInfo[] AllInfos = ReadInfos();

		static string AddSuffix(string s, char[] ca) =>
			string.Intern(s + new string(ca));

		static string AddPrefix(string s, char[] ca) =>
			string.Intern(new string(ca) + s);

		static InstrInfo[] ReadInfos() {
			var reader = new DataReader(GetSerializedInstrInfos());
			var infos = new InstrInfo[IcedConstants.CodeEnumCount];
			var strings = FormatterStringsTable.GetStringsTable();

			var ca = new char[1];
			string s, s2, s3, s4;
			uint v, v2, v3;
			int prevIndex = -1;
			for (int i = 0; i < infos.Length; i++) {
				byte f = reader.ReadByte();
				var ctorKind = (CtorKind)(f & 0x7F);
				int currentIndex;
				if (ctorKind == CtorKind.Previous) {
					currentIndex = reader.Index;
					reader.Index = prevIndex;
					ctorKind = (CtorKind)(reader.ReadByte() & 0x7F);
				}
				else {
					currentIndex = -1;
					prevIndex = reader.Index - 1;
				}
				s = strings[reader.ReadCompressedUInt32()];
				if ((f & 0x80) != 0) {
					ca[0] = 'v';
					s = AddPrefix(s, ca);
				}
				InstrInfo instrInfo;
				switch (ctorKind) {
				case CtorKind.Normal_1:
					instrInfo = new SimpleInstrInfo(s);
					break;

				case CtorKind.Normal_2:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo(s, (InstrOpInfoFlags)v);
					break;

				case CtorKind.AamAad:
					instrInfo = new SimpleInstrInfo_AamAad(s);
					break;

				case CtorKind.asz:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_as((int)v, s);
					break;

				case CtorKind.String:
					instrInfo = new SimpleInstrInfo_String(s);
					break;

				case CtorKind.bcst:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_bcst(s, (InstrOpInfoFlags)v);
					break;

				case CtorKind.bnd:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_bnd(s, (InstrOpInfoFlags)v);
					break;

				case CtorKind.DeclareData:
					instrInfo = new SimpleInstrInfo_DeclareData((Code)i, s);
					break;

				case CtorKind.er_2:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_er((int)v, s);
					break;

				case CtorKind.er_3:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_er((int)v, s, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.far:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_far((int)v, s);
					break;

				case CtorKind.far_mem:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_far_mem((int)v, s);
					break;

				case CtorKind.invlpga:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_invlpga((int)v, s);
					break;

				case CtorKind.maskmovq:
					instrInfo = new SimpleInstrInfo_maskmovq(s);
					break;

				case CtorKind.movabs:
					instrInfo = new SimpleInstrInfo_movabs(s);
					break;

				case CtorKind.nop:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_nop((int)v, s, (Register)v2);
					break;

				case CtorKind.OpSize:
					v = reader.ReadByte();
					ca[0] = 'w';
					s2 = AddSuffix(s, ca);
					ca[0] = 'd';
					s3 = AddSuffix(s, ca);
					ca[0] = 'q';
					s4 = AddSuffix(s, ca);
					instrInfo = new SimpleInstrInfo_OpSize((CodeSize)v, s, s2, s3, s4);
					break;

				case CtorKind.OpSize2_bnd:
					s2 = strings[reader.ReadCompressedUInt32()];
					s3 = strings[reader.ReadCompressedUInt32()];
					s4 = strings[reader.ReadCompressedUInt32()];
					instrInfo = new SimpleInstrInfo_OpSize2_bnd(s, s2, s3, s4);
					break;

				case CtorKind.OpSize3:
					ca[0] = (char)reader.ReadByte();
					s2 = AddSuffix(s, ca);
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_OpSize3((int)v, s, s2);
					break;

				case CtorKind.os_2:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os((int)v, s);
					break;

				case CtorKind.os_3:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os((int)v, s, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.os_call:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadByte();
					if (v2 > 1)
						throw new InvalidOperationException();
					instrInfo = new SimpleInstrInfo_os_call((int)v, s, v2 != 0);
					break;

				case CtorKind.CC_1:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_cc((int)v, new[] { s });
					break;

				case CtorKind.CC_2:
					s2 = strings[reader.ReadCompressedUInt32()];
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_cc((int)v, new[] { s, s2 });
					break;

				case CtorKind.CC_3:
					s2 = strings[reader.ReadCompressedUInt32()];
					s3 = strings[reader.ReadCompressedUInt32()];
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_cc((int)v, new[] { s, s2, s3 });
					break;

				case CtorKind.os_jcc_a_1:
					v2 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_jcc((int)v, (int)v2, new[] { s });
					break;

				case CtorKind.os_jcc_a_2:
					s2 = strings[reader.ReadCompressedUInt32()];
					v2 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_jcc((int)v, (int)v2, new[] { s, s2 });
					break;

				case CtorKind.os_jcc_a_3:
					s2 = strings[reader.ReadCompressedUInt32()];
					s3 = strings[reader.ReadCompressedUInt32()];
					v2 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_jcc((int)v, (int)v2, new[] { s, s2, s3 });
					break;

				case CtorKind.os_jcc_b_1:
					v3 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_jcc((int)v, (int)v3, new[] { s }, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.os_jcc_b_2:
					s2 = strings[reader.ReadCompressedUInt32()];
					v3 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_jcc((int)v, (int)v3, new[] { s, s2 }, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.os_jcc_b_3:
					s2 = strings[reader.ReadCompressedUInt32()];
					s3 = strings[reader.ReadCompressedUInt32()];
					v3 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_jcc((int)v, (int)v3, new[] { s, s2, s3 }, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.os_loopcc:
					s2 = strings[reader.ReadCompressedUInt32()];
					v3 = reader.ReadCompressedUInt32();
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_os_loop((int)v, (int)v3, (Register)v2, new[] { s, s2 });
					break;

				case CtorKind.os_loop:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_os_loop((int)v, -1, (Register)v2, new[] { s });
					break;

				case CtorKind.os_mem:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_mem((int)v, s);
					break;

				case CtorKind.os_mem_reg16:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_mem_reg16((int)v, s);
					break;

				case CtorKind.os_mem2:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_os_mem2((int)v, s, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.pblendvb:
					v = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_pblendvb(s, (MemorySize)v);
					break;

				case CtorKind.pclmulqdq:
					v = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_pclmulqdq(s, FormatterConstants.GetPseudoOps((PseudoOpsKind)v));
					break;

				case CtorKind.pops:
					v = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_pops(s, FormatterConstants.GetPseudoOps((PseudoOpsKind)v));
					break;

				case CtorKind.Reg16:
					instrInfo = new SimpleInstrInfo_Reg16(s);
					break;

				case CtorKind.Reg32:
					instrInfo = new SimpleInstrInfo_Reg32(s);
					break;

				case CtorKind.reverse:
					instrInfo = new SimpleInstrInfo_reverse(s);
					break;

				case CtorKind.sae:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_sae((int)v, s);
					break;

				case CtorKind.push_imm8:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_push_imm8((int)v, (SignExtendInfo)v2, s);
					break;

				case CtorKind.push_imm:
					v = reader.ReadCompressedUInt32();
					v2 = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_push_imm((int)v, (SignExtendInfo)v2, s);
					break;

				case CtorKind.SignExt_3:
					v = reader.ReadByte();
					v2 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_SignExt((SignExtendInfo)v, s, (InstrOpInfoFlags)v2);
					break;

				case CtorKind.SignExt_4:
					v = reader.ReadByte();
					v2 = reader.ReadByte();
					v3 = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_SignExt((SignExtendInfo)v, (SignExtendInfo)v2, s, (InstrOpInfoFlags)v3);
					break;

				case CtorKind.imul:
					v = reader.ReadByte();
					instrInfo = new SimpleInstrInfo_imul((SignExtendInfo)v, s);
					break;

				case CtorKind.STIG1:
					v = reader.ReadByte();
					if (v > 1)
						throw new InvalidOperationException();
					instrInfo = new SimpleInstrInfo_STIG1(s, v != 0);
					break;

				case CtorKind.STIG2_2a:
					v = reader.ReadByte();
					if (v > 1)
						throw new InvalidOperationException();
					instrInfo = new SimpleInstrInfo_STIG2(s, v != 0);
					break;

				case CtorKind.STIG2_2b:
					v = reader.ReadCompressedUInt32();
					instrInfo = new SimpleInstrInfo_STIG2(s, (InstrOpInfoFlags)v);
					break;

				case CtorKind.XLAT:
					instrInfo = new SimpleInstrInfo_XLAT(s);
					break;

				default:
					throw new InvalidOperationException();
				}
				infos[i] = instrInfo;
				if (currentIndex >= 0)
					reader.Index = currentIndex;
			}
			if (reader.CanRead)
				throw new InvalidOperationException();

			return infos;
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/fmt/nasm/InstrInfos.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt.nasm;

import com.github.icedland.iced.x86.internal.IcedConstants;
import com.github.icedland.iced.x86.internal.ResourceReader;
import com.github.icedland.iced.x86.internal.fmt.FormatterStringsTable;

final class InstrInfos {
	static final InstrInfo[] allInfos = readInfos();

	private static byte[] getSerializedInstrInfos() {
		return ResourceReader.readByteArray(InstrInfos.class.getClassLoader(), "com/github/icedland/iced/x86/fmt/nasm/InstrInfos.bin");
	}

	private static String addSuffix(String s, char[] ca) {
		return (s + new String(ca)).intern();
	}

	private static String addPrefix(String s, char[] ca) {
		return (new String(ca) + s).intern();
	}

	@SuppressWarnings("deprecation")
	private static InstrInfo[] readInfos() {
		com.github.icedland.iced.x86.internal.DataReader reader = new com.github.icedland.iced.x86.internal.DataReader(getSerializedInstrInfos());
		InstrInfo[] infos = new InstrInfo[IcedConstants.CODE_ENUM_COUNT];
		String[] strings = FormatterStringsTable.getStringsTable();

		char[] ca = new char[1];
		String s, s2, s3, s4;
		int v, v2, v3;
		int prevIndex = -1;
		for (int i = 0; i < infos.length; i++) {
			byte f = (byte)reader.readByte();
			byte ctorKind = (byte)(f & 0x7F);
			int currentIndex;
			if (ctorKind == CtorKind.PREVIOUS) {
				currentIndex = reader.getIndex();
				reader.setIndex(prevIndex);
				ctorKind = (byte)(reader.readByte() & 0x7F);
			}
			else {
				currentIndex = -1;
				prevIndex = reader.getIndex() - 1;
			}
			s = strings[reader.readCompressedUInt32()];
			if ((f & 0x80) != 0) {
				ca[0] = 'v';
				s = addPrefix(s, ca);
			}
			InstrInfo instrInfo;
			switch (ctorKind) {
			case CtorKind.NORMAL_1:
				instrInfo = new SimpleInstrInfo(s);
				break;

			case CtorKind.NORMAL_2:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo(s, v);
				break;

			case CtorKind.AAM_AAD:
				instrInfo = new SimpleInstrInfo_AamAad(s);
				break;

			case CtorKind.ASZ:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_as(v, s);
				break;

			case CtorKind.STRING:
				instrInfo = new SimpleInstrInfo_String(s);
				break;

			case CtorKind.BCST:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_bcst(s, v);
				break;

			case CtorKind.BND:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_bnd(s, v);
				break;

			case CtorKind.DECLARE_DATA:
				instrInfo = new SimpleInstrInfo_DeclareData(i, s);
				break;

			case CtorKind.ER_2:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_er(v, s);
				break;

			case CtorKind.ER_3:
				v = reader.readCompressedUInt32();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_er(v, s, v2);
				break;

			case CtorKind.FAR:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_far(v, s);
				break;

			case CtorKind.FAR_MEM:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_far_mem(v, s);
				break;

			case CtorKind.INVLPGA:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_invlpga(v, s);
				break;

			case CtorKind.MASKMOVQ:
				instrInfo = new SimpleInstrInfo_maskmovq(s);
				break;

			case CtorKind.MOVABS:
				instrInfo = new SimpleInstrInfo_movabs(s);
				break;

			case CtorKind.NOP:
				v = reader.readCompressedUInt32();
				v2 = reader.readByte();
				instrInfo = new SimpleInstrInfo_nop(v, s, v2);
				break;

			case CtorKind.OP_SIZE:
				v = reader.readByte();
				ca[0] = 'w';
				s2 = addSuffix(s, ca);
				ca[0] = 'd';
				s3 = addSuffix(s, ca);
				ca[0] = 'q';
				s4 = addSuffix(s, ca);
				instrInfo = new SimpleInstrInfo_OpSize(v, s, s2, s3, s4);
				break;

			case CtorKind.OP_SIZE2_BND:
				s2 = strings[reader.readCompressedUInt32()];
				s3 = strings[reader.readCompressedUInt32()];
				s4 = strings[reader.readCompressedUInt32()];
				instrInfo = new SimpleInstrInfo_OpSize2_bnd(s, s2, s3, s4);
				break;

			case CtorKind.OP_SIZE3:
				ca[0] = (char)reader.readByte();
				s2 = addSuffix(s, ca);
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_OpSize3(v, s, s2);
				break;

			case CtorKind.OS_2:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os(v, s);
				break;

			case CtorKind.OS_3:
				v = reader.readCompressedUInt32();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os(v, s, v2);
				break;

			case CtorKind.OS_CALL:
				v = reader.readCompressedUInt32();
				v2 = reader.readByte();
				if (v2 > 1)
					throw new UnsupportedOperationException();
				instrInfo = new SimpleInstrInfo_os_call(v, s, v2 != 0);
				break;

			case CtorKind.CC_1:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_cc(v, new String[] { s });
				break;

			case CtorKind.CC_2:
				s2 = strings[reader.readCompressedUInt32()];
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_cc(v, new String[] { s, s2 });
				break;

			case CtorKind.CC_3:
				s2 = strings[reader.readCompressedUInt32()];
				s3 = strings[reader.readCompressedUInt32()];
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_cc(v, new String[] { s, s2, s3 });
				break;

			case CtorKind.OS_JCC_A_1:
				v2 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_jcc(v, v2, new String[] { s });
				break;

			case CtorKind.OS_JCC_A_2:
				s2 = strings[reader.readCompressedUInt32()];
				v2 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_jcc(v, v2, new String[] { s, s2 });
				break;

			case CtorKind.OS_JCC_A_3:
				s2 = strings[reader.readCompressedUInt32()];
				s3 = strings[reader.readCompressedUInt32()];
				v2 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_jcc(v, v2, new String[] { s, s2, s3 });
				break;

			case CtorKind.OS_JCC_B_1:
				v3 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_jcc(v, v3, new String[] { s }, v2);
				break;

			case CtorKind.OS_JCC_B_2:
				s2 = strings[reader.readCompressedUInt32()];
				v3 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_jcc(v, v3, new String[] { s, s2 }, v2);
				break;

			case CtorKind.OS_JCC_B_3:
				s2 = strings[reader.readCompressedUInt32()];
				s3 = strings[reader.readCompressedUInt32()];
				v3 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_jcc(v, v3, new String[] { s, s2, s3 }, v2);
				break;

			case CtorKind.OS_LOOPCC:
				s2 = strings[reader.readCompressedUInt32()];
				v3 = reader.readCompressedUInt32();
				v = reader.readCompressedUInt32();
				v2 = reader.readByte();
				instrInfo = new SimpleInstrInfo_os_loop(v, v3, v2, new String[] { s, s2 });
				break;

			case CtorKind.OS_LOOP:
				v = reader.readCompressedUInt32();
				v2 = reader.readByte();
				instrInfo = new SimpleInstrInfo_os_loop(v, -1, v2, new String[] { s });
				break;

			case CtorKind.OS_MEM:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_mem(v, s);
				break;

			case CtorKind.OS_MEM_REG16:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_mem_reg16(v, s);
				break;

			case CtorKind.OS_MEM2:
				v = reader.readCompressedUInt32();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_os_mem2(v, s, v2);
				break;

			case CtorKind.PBLENDVB:
				v = reader.readByte();
				instrInfo = new SimpleInstrInfo_pblendvb(s, v);
				break;

			case CtorKind.PCLMULQDQ:
				v = reader.readByte();
				instrInfo = new SimpleInstrInfo_pclmulqdq(s, com.github.icedland.iced.x86.internal.fmt.FormatterConstants.getPseudoOps(v));
				break;

			case CtorKind.POPS:
				v = reader.readByte();
				instrInfo = new SimpleInstrInfo_pops(s, com.github.icedland.iced.x86.internal.fmt.FormatterConstants.getPseudoOps(v));
				break;

			case CtorKind.REG16:
				instrInfo = new SimpleInstrInfo_Reg16(s);
				break;

			case CtorKind.REG32:
				instrInfo = new SimpleInstrInfo_Reg32(s);
				break;

			case CtorKind.REVERSE:
				instrInfo = new SimpleInstrInfo_reverse(s);
				break;

			case CtorKind.SAE:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_sae(v, s);
				break;

			case CtorKind.PUSH_IMM8:
				v = reader.readCompressedUInt32();
				v2 = reader.readByte();
				instrInfo = new SimpleInstrInfo_push_imm8(v, v2, s);
				break;

			case CtorKind.PUSH_IMM:
				v = reader.readCompressedUInt32();
				v2 = reader.readByte();
				instrInfo = new SimpleInstrInfo_push_imm(v, v2, s);
				break;

			case CtorKind.SIGN_EXT_3:
				v = reader.readByte();
				v2 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_SignExt(v, s, v2);
				break;

			case CtorKind.SIGN_EXT_4:
				v = reader.readByte();
				v2 = reader.readByte();
				v3 = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_SignExt(v, v2, s, v3);
				break;

			case CtorKind.IMUL:
				v = reader.readByte();
				instrInfo = new SimpleInstrInfo_imul(v, s);
				break;

			case CtorKind.STIG1:
				v = reader.readByte();
				if (v > 1)
					throw new UnsupportedOperationException();
				instrInfo = new SimpleInstrInfo_STIG1(s, v != 0);
				break;

			case CtorKind.STIG2_2A:
				v = reader.readByte();
				if (v > 1)
					throw new UnsupportedOperationException();
				instrInfo = new SimpleInstrInfo_STIG2(s, v != 0);
				break;

			case CtorKind.STIG2_2B:
				v = reader.readCompressedUInt32();
				instrInfo = new SimpleInstrInfo_STIG2(s, v);
				break;

			case CtorKind.XLAT:
				instrInfo = new SimpleInstrInfo_XLAT(s);
				break;

			default:
				throw new UnsupportedOperationException();
			}
			infos[i] = instrInfo;
			if (currentIndex >= 0)
				reader.setIndex(currentIndex);
		}
		if (reader.canRead())
			throw new UnsupportedOperationException();

		return infos;
	}
}
